### PR TITLE
feat: implement KZG Params

### PR DESCRIPTION
**Pull request by Insun35**
# Description

- KZG Params
  - `k`: Degree of the polynomial
  - `n`: 2^k
  - `g1_powers_of_tau`: [𝜏⁰g₁, 𝜏¹g₁, ... , 𝜏ⁿ⁻¹g₁]
  - `g1_powers_of_tau_lagrange`: Bases multiplied by Lagrange coefficients of tau and g1
  - `g2`: G2 Generator
  - `tau_g2`:  𝜏g₂
- UnsafeSetup: Initialize parameters with a random toxic point.
- Commit: Commit to a polynomial with given coefficients.
- CommitLagrange: Commit to a polynomial with Lagrange coefficients.

This PR should be merged after #126

### CODE REVIEW

**Comment by chokobole** [comment]
`chore(crypto): fix typo` -> `build(crypto): fix typo`: See https://github.com/kroma-network/.github/blob/main/CONTRIBUTING.md#commit-type

File: tachyon/math/polynomials/univariate/univariate_dense_coefficients.h [hunk 1]
```diff
@@ -138,6 +138,8 @@ class UnivariateDenseCoefficients {
     return ss.str();
   }
 
+  std::vector<F> GetCoefficients() const { return coefficients_; }
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_dense_coefficients.h:141]: Please just make a getter and do not return it as a copy.

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  using DenseCoeffs = typename UnivariateEvaluationDomainTy::DenseCoeffs;
+  using G1JacobianPoint = typename G1PointTy::JacobianPointTy;
+
+  KZGParams() = default;
+
+  KZGParams(size_t k, std::vector<G1PointTy>&& g1_powers_of_tau,
+            std::vector<G1PointTy>&& g1_powers_of_tau_lagrange,
+            G2PointTy tau_g2)
+      : k_(k),
+        n_(1 << k),
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:37]:
```suggestion
        n_(size_t{1} << k),
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        g1_powers_of_tau_(std::move(g1_powers_of_tau)),
+        g1_powers_of_tau_lagrange_(std::move(g1_powers_of_tau_lagrange)),
+        tau_g2_(tau_g2) {}
+
+  size_t GetDegree() const { return k_; }
+
+  size_t GetSize() const { return n_; }
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:44]: Is there a reason for naming it as `GetDegree()` and `GetSize()`?

```suggestion
  size_t k() const { return k_; }

  size_t n() const { return n_; }
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  const std::vector<G1PointTy>& GetG1PowersOfTau() const {
+    return g1_powers_of_tau_;
+  }
+
+  const std::vector<G1PointTy>& GetG1PowersOfTauLagrange() const {
+    return g1_powers_of_tau_lagrange_;
+  }
+
+  const G2PointTy& GetTauG2() const { return tau_g2_; }
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:54]: Please name them as a getter

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Initialize parameters with a random toxic point.
+  // MUST NOT be used in production.
+  void UnsafeSetup(size_t degree) {
+    Fr tau = Fr::Random();
+    UnsafeSetupWithTau(degree, tau);
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:60]:
```suggestion
    UnsafeSetupWithTau(degree, Fr::Random());
```

File: tachyon/crypto/commitments/kzg/kzg_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#include "tachyon/math/elliptic_curves/bn/bn254/g2.h"
+#include "tachyon/math/polynomials/univariate/radix2_evaluation_domain.h"
+
+namespace tachyon::crypto {
+
+namespace {
+
+class KZGTest : public testing::Test {
+ public:
+  using UnivariateEvaluationDomainTy =
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/kzg/kzg_unittest.cc:15]: Please create the evaluation domain through a factory. See, univariate_evaluation_domain_factory.h.

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+#include "tachyon/math/polynomials/univariate/univariate_polynomial.h"
+
+namespace tachyon::crypto {
+
+template <typename G1PointTy, typename G2PointTy>
+class KZGParams {
+ public:
+  using Fr = typename G1PointTy::ScalarField;
+  using Bucket = typename math::Pippenger<G1PointTy>::Bucket;
+  using UnivariateEvaluationDomainTy = math::Radix2EvaluationDomain<Fr>;
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/kzg/kzg.h:27]: We decided to create the evaluation domain through a factory. See, univariate_evaluation_domain_factory.h. And the factory creates the `UnivariateEvaluationDomain` (not `Radix2EvaluationDomain`).

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/elliptic_curves/point_conversions.h [hunk 8]
```diff
@@ -14,6 +15,19 @@ constexpr DstPointTy ConvertPoint(const SrcPointTy& src_point) {
   return PointConversions<SrcPointTy, DstPointTy>::Convert(src_point);
 }
 
+// TODO(insun35): Parallelize ConvertPoints using OpenMP.
+template <typename DstContainer, typename SrcContainer>
+constexpr void ConvertPoints(const SrcContainer& src_points,
+                             DstContainer* dst_points) {
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/point_conversions.h:21]:
```suggestion
constexpr bool ConvertPoints(const SrcContainer& src_points,
                             DstContainer* dst_points) {
```

File: tachyon/math/elliptic_curves/point_conversions.h [hunk 9]
```diff
@@ -14,6 +15,19 @@ constexpr DstPointTy ConvertPoint(const SrcPointTy& src_point) {
...
+  using SrcPointTy = typename SrcContainer::value_type;
+  using DstPointTy = typename DstContainer::value_type;
+
+  dst_points->resize(src_points.size());
+  for (size_t i = 0; i < src_points.size(); ++i) {
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/point_conversions.h:26]:
```suggestion
  if (std::size(src_points) != std::size(dst_points)) return false;
  for (size_t i = 0; i < std::size(src_points); ++i) {
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+  KZGParams() = default;
+
+  KZGParams(size_t k, std::vector<G1PointTy> g1_powers_of_tau,
+            std::vector<G1PointTy> g1_powers_of_tau_lagrange, G2PointTy tau_g2)
+      : k_(k),
+        n_((base::CheckedNumeric<size_t>(1) << k).ValueOrDie()),
+        g1_powers_of_tau_(std::move(g1_powers_of_tau)),
+        g1_powers_of_tau_lagrange_(std::move(g1_powers_of_tau_lagrange)),
+        tau_g2_(std::move(tau_g2)) {}
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:36]:
```suggestion
        tau_g2_(std::move(tau_g2)) {
          CHECK_LE(k, n);
        }
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    std::vector<typename G1PointTy::JacobianPointTy> g1_powers_of_tau_jacobian;
+
+    g1_powers_of_tau_jacobian.resize(n_);
+    if (!G1PointTy::MultiScalarMul(powers_of_tau, g1,
+                                   &g1_powers_of_tau_jacobian)) {
+      return false;
+    }
+
+    math::ConvertPoints<std::vector<G1PointTy>>(g1_powers_of_tau_jacobian,
+                                                &g1_powers_of_tau_);
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:77]:
```suggestion
    if (!math::ConvertPoints<std::vector<G1PointTy>>(g1_powers_of_tau_jacobian,
                                                &g1_powers_of_tau_)) { return false; }
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        g1_powers_of_tau_lagrange_jacobian;
+
+    g1_powers_of_tau_lagrange_jacobian.resize(n_);
+    if (!G1PointTy::MultiScalarMul(lagrange_coeffs.coefficients(), g1,
+                                   &g1_powers_of_tau_lagrange_jacobian)) {
+      return false;
+    }
+
+    math::ConvertPoints<std::vector<G1PointTy>>(
+        g1_powers_of_tau_lagrange_jacobian, &g1_powers_of_tau_lagrange_);
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:94]:
```suggestion
    if (!math::ConvertPoints<std::vector<G1PointTy>>(
        g1_powers_of_tau_lagrange_jacobian, &g1_powers_of_tau_lagrange_)) { return false; }
```

File: tachyon/crypto/commitments/kzg/kzg_unittest.cc [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      domain->FFT(poly);
+
+  math::bn254::G1AffinePoint commit_lagrange;
+  ASSERT_TRUE(kzg.CommitLagrange(poly_evals, &commit_lagrange));
+
+  EXPECT_EQ(commit, commit_lagrange);
+
+  kzg.Downsize(3);
+  EXPECT_EQ(kzg.n(), size_t{8});
+}
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg_unittest.cc:45]: This test are mixed up not only `UnsafeSetup` but also various things. Could you split them?

**Review by chokobole — CHANGES_REQUESTED** [review]

**Review by chokobole — APPROVED** [review]
LGTM

File: tachyon/crypto/commitments/kzg/kzg_unittest.cc [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#include "tachyon/crypto/commitments/kzg/kzg.h"
+
+#include "gtest/gtest.h"
+
+#include "tachyon/base/numerics/checked_math.h"
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg_unittest.cc:5]: Please remove this.

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        g1_powers_of_tau_lagrange_jacobian;
+
+    g1_powers_of_tau_lagrange_jacobian.resize(n_);
+    if (!G1PointTy::MultiScalarMul(lagrange_coeffs.coefficients(), g1,
+                                   &g1_powers_of_tau_lagrange_jacobian))
+      return false;
+
+    g1_powers_of_tau_lagrange_.resize(n_);
+    if (!math::ConvertPoints<std::vector<G1PointTy>>(
+            g1_powers_of_tau_lagrange_jacobian, &g1_powers_of_tau_lagrange_)) {
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:98]:
```suggestion
    if (!math::ConvertPoints(g1_powers_of_tau_lagrange_jacobian, &g1_powers_of_tau_lagrange_)) {
```

**Review by chokobole — CHANGES_REQUESTED** [review]

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+  [[nodiscard]] bool DoCommit(const Poly& poly,
+                              const std::vector<G1PointTy>& bases,
+                              G1PointTy* out) const {
+    math::VariableBaseMSM<G1PointTy> msm;
+    typename math::Pippenger<G1PointTy>::Bucket bucket;
+    if (!msm.Run(bases, poly, &bucket)) return false;
+
+    *out = math::ConvertPoint<G1PointTy>(bucket);
+    return true;
+  }
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:148]: Please move members down

File: tachyon/math/polynomials/multivariate/multivariate_sparse_coefficients.h [hunk 17]
```diff
@@ -329,6 +329,10 @@ class MultivariateSparseCoefficients {
                   [](const Term& term) { return term.coefficient.IsZero(); });
   }
 
+  const size_t num_vars() const { return num_vars_; }
```
Comment (chokobole) [inline tachyon/math/polynomials/multivariate/multivariate_sparse_coefficients.h:332]:
```suggestion
  size_t num_vars() const { return num_vars_; }
```

**Review by chokobole — APPROVED** [review]
LGTM

**Pull request merged**